File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.go
func maxProfit(prices []int) int {
    if len(prices) < 2 {
        return 0
    }

	maxProfit := 0
    currentBuy := prices[0]
    currentSell := prices[1]
    
    for i := 0; i < len(prices); i++ {

        if prices[i] < currentBuy {
            if i+1 < len(prices) {
                currentBuy = prices[i]
                currentSell = prices[i+1]

                if (currentSell - currentBuy) > maxProfit {
                    maxProfit = currentSell - currentBuy
                }
            }
        } 

        if prices[i] > currentBuy {
            currentSell = prices[i]
            if (currentSell - currentBuy) > maxProfit {
                maxProfit = currentSell - currentBuy
            }
        } 
    
    } 

    return maxProfit

}

File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.js
/**
 * @param {number[]} prices
 * @return {number}
 */
var maxProfit = function(prices) {
    let buy = prices[0];
    let res = 0;
    for(let i = 1; i < prices.length; i++){
        if(buy > prices[i]){
            buy = prices[i];
        }
        let profit = prices[i] - buy;
        res = Math.max(res, profit);
    }  

    return res
};